File: packages/agent/src/grammar/entity-candidates.ts
import { Registry, Entity } from './registry.js';

// Score how well an entity matches the question.
//
// Beyond the previous full-name matching, we now also tokenize the entity
// name on '_' and score each token against the question. This handles cases
// like Django's `projects_project` vs `projects_projecttemplate`:
//   Q: "how many projects"
//   - `projects_project` token "project" matches "projects" → token score
//   - `projects_projecttemplate` token "projecttemplate" → no match
//
// Tie-breakers (in order):
//   1. Higher score wins
//   2. Fewer name segments wins (`project` over `projects_project`,
//      `projects_project` over `projects_projecttemplate`)
//   3. Higher row count (data table over lookup/template table)
function scoreEntity(question: string, entity: Entity, registry: Registry): number {
  const q = question.toLowerCase();
  const singular = entity.name.toLowerCase();
  const plural = entity.table.toLowerCase();
  let score = 0;

  // Whole-name word boundary match (highest signal)
  if (q.includes(` ${singular} `) || q.startsWith(`${singular} `) || q.endsWith(` ${singular}`)) score += 12;
  if (q.includes(plural)) score += 10;
  if (q.includes(singular)) score += 5;

  // Alias match
  for (const [alias, target] of Object.entries(registry.aliases)) {
    if (target !== entity.name) continue;
    if (q.includes(alias.toLowerCase())) score += 8;
  }

  // Token-level matching: split entity name on '_' and score each token.
  // This catches `projects_project` → "project" matches "projects".
  const tokens = singular.split('_').filter(t => t.length >= 3);
  for (const tok of tokens) {
    const tokPlural = pluralizeSimple(tok);
    // Whole-word match against the question
    const re = new RegExp(`\\b(${escapeRegex(tok)}|${escapeRegex(tokPlural)})\\b`);
    if (re.test(q)) score += 4;
  }

  // Whitespace-collapsed match: "user stories" → "userstories" → matches
  // "userstories_userstory" token. Score is length-weighted so longer
  // matches win (avoids "user" in "userstories" giving spurious credit
  // to a `users_user` table when the real intent is `userstories_userstory`).
  const qCompact = q.replace(/\s+/g, '');
  let bestCompactLen = 0;
  for (const tok of tokens) {
    if (tok.length < 5) continue;
    const candidates = [tok, pluralizeSimple(tok)];
    for (const c of candidates) {
      if (qCompact.includes(c) && c.length > bestCompactLen) bestCompactLen = c.length;
    }
  }
  if (bestCompactLen > 0) score += bestCompactLen; // 5..15+

  return score;
}

function pluralizeSimple(word: string): string {
  if (/(s|x|ch|sh)$/.test(word)) return word + 'es';
  if (/[^aeiou]y$/.test(word)) return word.slice(0, -1) + 'ies';
  return word + 's';
}

function escapeRegex(s: string): string {
  return s.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

// Number of segments in entity name. Simpler names get tie-break preference.
function nameSegments(entity: Entity): number {
  return entity.name.split('_').length;
}

export function selectEntityCandidates(question: string, registry: Registry, topN: number): Entity[] {
  const scored = Object.values(registry.entities).map(entity => ({
    entity,
    score: scoreEntity(question, entity, registry),
    segments: nameSegments(entity),
    rowCount: entity.rowCount,
  }));

  scored.sort((a, b) => {
    if (b.score !== a.score) return b.score - a.score;            // higher score first
    if (a.segments !== b.segments) return a.segments - b.segments; // simpler name first
    return b.rowCount - a.rowCount;                                // more rows last (real data > lookup)
  });

  if (scored[0]?.score === 0) {
    // No term matched anywhere. Fall back to high-row-count entities.
    return Object.values(registry.entities)
      .sort((a, b) => b.rowCount - a.rowCount)
      .slice(0, topN);
  }
  return scored.slice(0, topN).map(s => s.entity);
}

File: sql-chatbot-rails/lib/sql_chatbot/grammar/entity_candidates.rb
# frozen_string_literal: true

require "sql_chatbot/grammar/registry"

module SqlChatbot
  module Grammar
    module EntityCandidates
      # Score an entity against the question.
      # Tokenizes name on '_' so e.g. `projects_project` token "project" matches
      # the question "how many projects". Tie-breakers: fewer name segments,
      # then higher row count.
      def self.score_entity(question, entity, registry)
        q = question.to_s.downcase
        singular = entity.name.to_s.downcase
        plural = entity.table.to_s.downcase
        score = 0

        score += 12 if q.include?(" #{singular} ") || q.start_with?("#{singular} ") || q.end_with?(" #{singular}")
        score += 10 if q.include?(plural)
        score += 5 if q.include?(singular)

        registry.aliases.each do |alias_term, target|
          next unless target == entity.name
          score += 8 if q.include?(alias_term.to_s.downcase)
        end

        # Token-level matching for compound names like `projects_project`.
        tokens = singular.split("_").select { |t| t.length >= 3 }
        tokens.each do |tok|
          tok_plural = pluralize_simple(tok)
          if q =~ /\b(#{Regexp.escape(tok)}|#{Regexp.escape(tok_plural)})\b/
            score += 4
          end
        end

        # Whitespace-collapsed match — length-weighted so longer matches win.
        q_compact = q.gsub(/\s+/, "")
        best_len = 0
        tokens.each do |tok|
          next if tok.length < 5
          [tok, pluralize_simple(tok)].each do |c|
            best_len = c.length if q_compact.include?(c) && c.length > best_len
          end
        end
        score += best_len

        score
      end

      def self.pluralize_simple(word)
        return word + "es" if word.end_with?("s", "x", "ch", "sh")
        return word[0..-2] + "ies" if word.end_with?("y") && !%w[a e i o u].include?(word[-2])
        word + "s"
      end

      def self.name_segments(entity)
        entity.name.to_s.split("_").length
      end

      def self.select(question:, registry:, top_n:)
        rows = registry.entities.values.map do |entity|
          {
            entity: entity,
            score: score_entity(question, entity, registry),
            segments: name_segments(entity),
            row_count: entity.row_count,
          }
        end

        rows.sort_by! { |r| [-r[:score], r[:segments], -r[:row_count]] }

        if rows.first && rows.first[:score] == 0
          return registry.entities.values.sort_by { |e| -e.row_count }.first(top_n)
        end
        rows.first(top_n).map { |r| r[:entity] }
      end
    end
  end
end
